lib/jellyfish/room.ex

```diff
@@ -0,0 +1,186 @@
+defmodule Jellyfish.Room do
+  @moduledoc """
+  Module representing room.
+
+  """
+
+  use Bunch.Access
+  use GenServer
+  alias Jellyfish.Peer
+  alias Jellyfish.Component
+  alias Membrane.RTC.Engine
+
+  @enforce_keys [
+    :id,
+    :config,
+    :engine_pid,
+    :network_options
+  ]
+  defstruct @enforce_keys ++ [components: %{}, peers: %{}]
+
+  @type id :: String.t()
+  @type max_peers :: integer() | nil
+
+  @typedoc """
+  This module contains:
+  * `id` - room id
+  * `config` - configuration of room. For example you can specify maximal number of peers
+  * `components` - list of components
+  * `peers` - list of peers
+  * `engine` - pid of engine
+  """
+  @type t :: %__MODULE__{
+          id: id,
+          config: %{max_peers: max_peers(), simulcast?: boolean()},
+          components: %{},
+          peers: %{Peer.id() => Peer.t()},
+          engine_pid: pid(),
+          network_options: %{}
+        }
+
+  @mix_env Mix.env()
+
+  def start(init_arg, opts) do
+    GenServer.start(__MODULE__, init_arg, opts)
+  end
+
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, [], opts)
+  end
+
+  @impl true
+  def init(max_peers) do
+    state = new(max_peers)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_call(:state, _from, state) do
+    engine_endpoints = Engine.get_endpoints(state.engine_pid)
+
+    peers =
+      engine_endpoints
+      |> Enum.filter(fn endpoint -> Map.has_key?(state.peers, endpoint.id) end)
+      |> Map.new(&{&1.id, &1})
+
+    components =
+      engine_endpoints
+      |> Enum.filter(fn endpoint ->
+        Map.has_key?(state.components, endpoint.id)
+      end)
+      |> Map.new(&{&1.id, &1})
+
+    {:reply, %{id: state.id, peers: peers, components: components, config: state.config}, state}
+  end
+
+  def handle_call({:add_peer, peer_type}, _from, state) do
+    if Enum.count(state.peers) == state.config.max_peers do
+      {:reply, {:error, :reached_peers_limit}, state}
+    else
+      options = %{engine_pid: state.engine_pid, network_options: state.network_options}
+      peer = Peer.create_peer(peer_type, options)
+
+      state = put_in(state, [:peers, peer.id], peer)
+
+      :ok = Engine.add_endpoint(state.engine_pid, peer.engine_endpoint, endpoint_id: peer.id)
```

Comment:
[As you can see](https://github.com/jellyfish-dev/membrane_rtc_engine/blob/8b6c2276bc88c7364769f92c51121948ed27a9db/lib/membrane_rtc_engine/engine.ex#L286-L298) always return `:ok` or raise an error, so it will never return `:error`